rewrite-maven/src/main/java/org/openrewrite/maven/RemoveUnusedProperties.java

```diff
@@ -0,0 +1,274 @@
+        return "Remove unused properties";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Detect and remove Maven property declarations which do not have any usage within the project.";
+    }
+
+    public static class Accumulator {
+        public Map<String, Set<MavenResolutionResult>> propertiesToUsingPoms = new HashMap<>();
+        public Map<Path, MavenResolutionResult> filteredResourcePathsToDeclaringPoms = new HashMap<>();
+        public Map<Path, Set<String>> nonPomPathsToUsages = new HashMap<>();
+
+        public Map<String, Set<MavenResolutionResult>> getFilteredResourceUsages() {
+            Map<String, Set<MavenResolutionResult>> result = new HashMap<>();
+            filteredResourcePathsToDeclaringPoms.forEach((filteredResourcePath, mrr) ->
+                    nonPomPathsToUsages.forEach((usagePath, properties) -> {
+                        if (usagePath.startsWith(filteredResourcePath)) {
+                            properties.forEach(property -> {
+                                result.putIfAbsent(property, new HashSet<>());
+                                result.get(property).add(mrr);
+                            });
+                        }
+                    }
+            ));
+            return result;
+        }
+    }
+
+    @Override
+    public RemoveUnusedProperties.Accumulator getInitialValue(ExecutionContext ctx) {
+        return new RemoveUnusedProperties.Accumulator();
+    }
+
+    private String getPropertyPattern() {
+        return propertyPattern != null ? propertyPattern : ".+?";
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getScanner(RemoveUnusedProperties.Accumulator acc) {
+        Pattern propertyMatcher = Pattern.compile(getPropertyPattern());
+        Pattern propertyUsageMatcher = Pattern.compile("[^$]*\\$\\{(" + propertyMatcher.pattern() + ")}[^$]*");
+
+        MavenIsoVisitor<ExecutionContext> findPomUsagesVisitor =
+                new FindPomUsagesVisitor(propertyUsageMatcher, acc);
+        MavenIsoVisitor<ExecutionContext> findFilteredResourcePathsVisitor =
+                new FindFilteredResourcePathsVisitor(acc);
+        PlainTextVisitor<ExecutionContext> findResourceUsagesVisitor =
+                new FindResourceUsagesVisitor(propertyUsageMatcher, acc);
+
+        return new TreeVisitor<Tree, ExecutionContext>() {
+            @Override
+            public @Nullable Tree visit(@Nullable Tree tree, ExecutionContext ctx) {
+                if (tree instanceof SourceFile) {
+                    SourceFile sf = (SourceFile) tree;
+                    if (findPomUsagesVisitor.isAcceptable(sf, ctx)) { // ie: is a pom
+                        findPomUsagesVisitor.visit(sf, ctx);
+                        findFilteredResourcePathsVisitor.visit(sf, ctx);
+                    } else if (!(tree instanceof JavaSourceFile)) { // optimization: avoid visiting code files which are almost always not filtered resources
```

Comment:
my first instinct was to check the `JavaSourceSet` marker for this, but I didn't see a way to differentiate between code files and resources with that marker